File: python/PIEstimation.py
from Pyro4 import expose
import random
import time

class Solver:
    def __init__(self, workers=None, input_file_name=None, output_file_name=None):
        self.input_file_name = input_file_name
        self.output_file_name = output_file_name
        self.workers = workers
        print("Initialized Pi Estimation Solver")

    def solve(self):
        print("Pi Estimation Started")
        print("Workers: %d" % len(self.workers))
        self.total_darts = self.read_input()

        start_time = time.time()
        estimated_pi = self.parallel_estimate_pi()
        end_time = time.time()

        execution_time = end_time - start_time
        self.write_output(execution_time, estimated_pi)

    def parallel_estimate_pi(self):
        darts_per_process = self.total_darts // len(self.workers)
                    
        futures = [self.workers[i].estimate_pi_worker(darts_per_process) for i in range(len(self.workers))]
        results = [future.value for future in futures]
        total_hits = sum(results)

        return 4 * (float(total_hits) / self.total_darts)

    @staticmethod
    @expose
    def estimate_pi_worker(darts_per_process):
        hits = 0
        for _ in range(darts_per_process):
            x = random.uniform(-1, 1)
            y = random.uniform(-1, 1)
            if x**2 + y**2 <= 1:
                hits += 1
        return hits
    
    def read_input(self):
        with open(self.input_file_name, 'r') as f:
            line = f.readline().strip()
            return int(line)


    def write_output(self, execution_time, estimated_pi):
        with open(self.output_file_name, 'a') as f:
            f.write('Total Darts: {}\n'.format(self.total_darts))
            f.write('Execution Time: {:.6f} seconds\n'.format(execution_time))
            f.write('Estimated Pi: {:.8f}\n'.format(estimated_pi))
        print("Pi estimation complete. Execution time and result written to file.")

File: java/src/PiEstimation.java
import java.io.*;
import java.util.*;
import parcs.*;

public class PiEstimation implements AM {

    private static long startTime = 0;

    public static void startTimer() {
        startTime = System.nanoTime();
    }

    public static void stopTimer() {
        long endTime = System.nanoTime();
        long timeElapsed = endTime - startTime;
        double seconds = timeElapsed / 1_000_000_000.0;
        System.err.println("Time passed: " + seconds + " seconds.");
    }

    public static void main(String[] args) throws Exception {
        if (args.length != 2) {
            System.err.println("Usage: PiEstimation <total-darts> <number-of-workers>");
            System.exit(1);
        }

        int totalDarts = Integer.parseInt(args[0]);
        int k = Integer.parseInt(args[1]);
        System.err.println("Here");

        task curtask = new task();
        curtask.addJarFile("PiEstimation.jar");
        AMInfo info = new AMInfo(curtask, null);

        System.err.println("Distributing work to workers...");
        startTimer();
        channel[] channels = new channel[k];
        int dartsPerWorker = totalDarts / k;

        for (int i = 0; i < k; i++) {
            point p = info.createPoint();
            channel c = p.createChannel();
            int[] darts = new int[1];
            darts[0] = dartsPerWorker;
            p.execute("PiEstimation");
            c.write(darts);
            channels[i] = c;
        }
        stopTimer();

        System.err.println("Collecting results from workers...");

        startTimer();
        int totalHits = 0;
        for (int i = 0; i < k; i++) {
            int newHints = ((int[]) channels[i].readObject())[0];
            totalHits += newHints;
        }
        stopTimer();

        double estimatedPi = 4.0 * totalHits / (double) totalDarts;
        System.out.println("Estimated Pi: " + estimatedPi);

        curtask.end();
    }

    public void run(AMInfo info) {
        int[] dartsPerWorker = (int[]) info.parent.readObject();
        int[]hits = new int[1];
        hits[0] = estimateHits(dartsPerWorker[0]);
        info.parent.write(hits);
    }

    private int estimateHits(int dartsPerWorker) {
        Random random = new Random();
        int hits = 0;
        for (int i = 0; i < dartsPerWorker; i++) {
            double x = random.nextDouble() * 2 - 1;
            double y = random.nextDouble() * 2 - 1;
            if (x * x + y * y <= 1) {
                hits++;
            }
        }
        return hits;
    }
}